src/Categories/Diagram/End/Parameterized.agda

```diff
@@ -0,0 +1,220 @@
+{-# OPTIONS --without-K --safe #-}
+module Categories.Diagram.End.Parameterized where
+
+open import Level
+open import Function using (_$_)
+open import Data.Product using (Σ; _,_)
+
+open import Categories.Category
+open import Categories.Category.Construction.Functors
+open import Categories.Category.Product renaming (Product to _×ᶜ_)
+open import Categories.Diagram.End renaming (End to ∫)
+open import Categories.Diagram.End.Properties
+open import Categories.Diagram.Wedge
+open import Categories.Functor hiding (id)
+open import Categories.Functor.Bifunctor
+open import Categories.Functor.Bifunctor.Properties
+open import Categories.Functor.Limits
+open import Categories.NaturalTransformation renaming (_∘ʳ_ to _▹ⁿ_; id to idN)
+open import Categories.NaturalTransformation.Dinatural hiding (_≃_)
+open import Categories.NaturalTransformation.NaturalIsomorphism renaming (_≃_ to _≃ⁱ_)
+open import Categories.NaturalTransformation.Equivalence
+
+import Categories.Category.Construction.Wedges as Wedges
+import Categories.Morphism as M
+import Categories.Morphism.Reasoning as MR
+
+-- The following conventions are taken in this file: C is the 'source' category
+-- and D is the destination. If two source categories are needed, the other is
+-- called 'P' for "parameter", following MacLane. F, G and H are functors and ef,
+-- eg and eh are witnesses of their respective ends.
+
+private
+  variable
+    o ℓ e : Level
+    P C D E : Category o ℓ e
+
+-- we use MacLane's notation for paramaterized ends (CWM §IX.5)
+_♯ : Functor (P ×ᶜ (Category.op C ×ᶜ C)) D → Functor (Category.op C ×ᶜ C) (Functors P D)
+F ♯ = curry.₀ F.flip
+  where module F = Bifunctor F
+
+end-η♯ : {F G : Functor (P ×ᶜ (Category.op C ×ᶜ C)) D} (η : NaturalTransformation F G)
+         ⦃ ef : ∫ (F ♯) ⦄ ⦃ eg : ∫ (G ♯) ⦄ → NaturalTransformation (∫.E ef) (∫.E eg)
```

Comment:
please don't use instance arguments. I'm fine with `opaque` and `unfolding`, but not these. Things are slow enough as it is, let's not make it worse.

Reply:
I'm not sure if the instances are making things slow here.

They make a lot of definitions nicer. For example, in `Properties.agda` there is the type

```
end-η-resp-∘ : (α : NaturalTransformation F G) (β : NaturalTransformation G H) {{ef : ∫ F}} {{eg : ∫ G}} {{eh : ∫ H}} → end-η (β ∘ᵥ α) ≈ end-η β ∘ end-η α
```

Stating each end here for `end-η` complicates the definition a lot, and I don't think that it makes things nicer. Using implicit arguments isn't possible either, since the only portion of the end that is solved uniquely is the apex.